cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -0,0 +1,379 @@
+        uniq_elem_size = [&]() -> size_type {
+          if (not is_unique) { return 0; }
+          switch (s_col.physical_type) {
+            case Type::INT32: return 4;
+            case Type::INT64: return 8;
+            case Type::INT96: return 12;
+            case Type::FLOAT: return 4;
+            case Type::DOUBLE: return 8;
+            case Type::BYTE_ARRAY:
+              if (data_col.type().id() == type_id::STRING) {
+                // Strings are stored as 4 byte length + string bytes
+                return 4 + data_col.element<string_view>(val_idx).size_bytes();
+              }
+            case Type::FIXED_LEN_BYTE_ARRAY:
+            default: cudf_assert(false && "Unsupported type for dictionary encoding"); return 0;
+          }
+        }();
+      }
+    }
+
+    // TODO: Explore using ballot and popc for this reduction as the value to reduce is 1 or 0
+    auto num_unique = block_reduce(reduce_storage).Sum(is_unique);
+    __syncthreads();
+    auto uniq_data_size = block_reduce(reduce_storage).Sum(uniq_elem_size);
+    if (t == 0) {
+      atomicAdd(&s_chunk->num_dict_entries, num_unique);
+      // TODO: move out of loop. We don't have to update other blocks about this
+      atomicAdd(&s_chunk->uniq_data_size, uniq_data_size);
+    }
+    __syncthreads();
+  }
+}
+
+template <int block_size>
+__global__ void __launch_bounds__(block_size, 1)
+  gpuCollectMapEntries(device_span<EncColumnChunk> chunks)
+{
+  auto &chunk = chunks[blockIdx.x];
+  if (not chunk.use_dictionary) { return; }
+
+  auto t = threadIdx.x;
+  auto map =
+    map_type::device_view(chunk.dict_map_slots, chunk.dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
+
+  // TODO: Does this need to be volatile?
+  __shared__ size_type counter;
+  if (t == 0) counter = 0;
+  __syncthreads();
+  for (size_t i = 0; i < chunk.dict_map_size; i += block_size) {
+    if (t + i < chunk.dict_map_size) {
+      auto slot = map.begin_slot() + t + i;
+      auto key  = static_cast<map_type::key_type>(slot->first);
+      if (key != KEY_SENTINEL) {
+        auto loc = atomicAdd(&counter, 1);
+        cudf_assert(loc < MAX_DICT_SIZE && "Number of filled slots exceeds max dict size");
+        chunk.dict_data[loc] = key;
+        // If sorting dict page ever becomes a hard requirement, enable the following statement and
+        // add a dict sorting step before storing into the slot's second field.
+        // chunk.dict_data_idx[loc] = t + i;
+        slot->second.store(loc);
+        // TODO: ^ This doesn't need to be atomic. Try casting to value_type ptr and just writing.
+      }
+    }
+  }
+}
+
+template <int block_size>
+__global__ void __launch_bounds__(block_size, 1)
+  gpuGetDictionaryIndices(cudf::detail::device_2dspan<EncColumnChunk> chunks, size_type num_rows)
+{
+  auto col_idx = blockIdx.y;
+  auto block_x = blockIdx.x;
+  auto t       = threadIdx.x;
+
+  size_type start_row = block_x * 5000;
+  size_type end_row   = min(start_row + 5000, num_rows);
```

Comment:
`min` maybe the C++ macro.
```suggestion
  size_type end_row   = std::min(start_row + 5000, num_rows);
```

Reply:
This will use a faster cuda intrinsic https://stackoverflow.com/a/29979109/3325146

Reply:
Then we need to explicitly add the prefix `cuda::min` or something similar to that?

Reply:
We don't for now. It just works. But perhaps they will be included in a subsequent libcudacxx version to replace `std::min`